// @mysten/enoki/dist/esm/EnokiFlow.js
var __accessCheck = (obj, member, msg) => {
  if (!member.has(obj))
    throw TypeError("Cannot " + msg);
};
var __privateGet = (obj, member, getter) => {
  __accessCheck(obj, member, "read from private field");
  return getter ? getter.call(obj) : member.get(obj);
};
var __privateAdd = (obj, member, value) => {
  if (member.has(obj))
    throw TypeError("Cannot add the same private member more than once");
  member instanceof WeakSet ? member.add(obj) : member.set(obj, value);
};
var __privateSet = (obj, member, value, setter) => {
  __accessCheck(obj, member, "write to private field");
  setter ? setter.call(obj, value) : member.set(obj, value);
  return value;
};
var __privateMethod = (obj, member, method) => {
  __accessCheck(obj, member, "access private method");
  return method;
};
var _storageKeys, _enokiClient, _encryption, _encryptionKey, _store, _setSession, setSession_fn;
import { decodeSuiPrivateKey } from "@mysten/sui.js/cryptography";
import { Ed25519Keypair } from "@mysten/sui.js/keypairs/ed25519";
import { fromB64, toB64 } from "@mysten/sui.js/utils";
import { decodeJwt } from "jose";
import { atom, onMount, onSet } from "nanostores";
import { createDefaultEncryption } from "./encryption.js";
import { EnokiClient } from "./EnokiClient/index.js";
import { EnokiKeypair } from "./EnokiKeypair.js";
import { createSessionStorage } from "./stores.js";
const createStorageKeys = (apiKey) => ({
  STATE: `@enoki/flow/state/${apiKey}`,
  SESSION: `@enoki/flow/session/${apiKey}`
});
class EnokiFlow {
  constructor(config) {
    __privateAdd(this, _setSession);
    __privateAdd(this, _storageKeys, void 0);
    __privateAdd(this, _enokiClient, void 0);
    __privateAdd(this, _encryption, void 0);
    __privateAdd(this, _encryptionKey, void 0);
    __privateAdd(this, _store, void 0);
    __privateSet(this, _enokiClient, new EnokiClient({
      apiKey: config.apiKey,
      apiUrl: config.apiUrl
    }));
    __privateSet(this, _encryptionKey, config.apiKey);
    __privateSet(this, _encryption, config.encryption ?? createDefaultEncryption());
    __privateSet(this, _store, config.store ?? createSessionStorage());
    __privateSet(this, _storageKeys, createStorageKeys(config.apiKey));
    let storedState = null;
    try {
      const rawStoredValue = __privateGet(this, _store).get(__privateGet(this, _storageKeys).STATE);
      if (rawStoredValue) {
        storedState = JSON.parse(rawStoredValue);
      }
    } catch {
    }
    this.$zkLoginState = atom(storedState || {});
    this.$zkLoginSession = atom({ initialized: false, value: null });
    onMount(this.$zkLoginSession, () => {
      this.getSession();
    });
    onSet(this.$zkLoginState, ({ newValue }) => {
      __privateGet(this, _store).set(__privateGet(this, _storageKeys).STATE, JSON.stringify(newValue));
    });
  }
  get enokiClient() {
    return __privateGet(this, _enokiClient);
  }
  async createAuthorizationURL(input) {
    const ephemeralKeyPair = new Ed25519Keypair();
    const { nonce, randomness, maxEpoch, estimatedExpiration } = await __privateGet(this, _enokiClient).createZkLoginNonce({
      ephemeralPublicKey: ephemeralKeyPair.getPublicKey()
    });
    const params = new URLSearchParams({
      ...input.extraParams,
      nonce,
      client_id: input.clientId,
      redirect_uri: input.redirectUrl,
      response_type: "id_token",
      // TODO: Eventually fetch the scopes for this client ID from the Enoki service:
      scope: [
        "openid",
        // Merge the requested scopes in with the required openid scopes:
        ...input.extraParams && "scope" in input.extraParams ? input.extraParams.scope : []
      ].filter(Boolean).join(" ")
    });
    let oauthUrl;
    switch (input.provider) {
      case "google": {
        oauthUrl = `https://accounts.google.com/o/oauth2/v2/auth?${params}`;
        break;
      }
      case "facebook": {
        oauthUrl = `https://www.facebook.com/v17.0/dialog/oauth?${params}`;
        break;
      }
      case "twitch": {
        params.set("force_verify", "true");
        oauthUrl = `https://id.twitch.tv/oauth2/authorize?${params}`;
        break;
      }
      default:
        throw new Error(`Invalid provider: ${input.provider}`);
    }
    this.$zkLoginState.set({ provider: input.provider });
    await __privateMethod(this, _setSession, setSession_fn).call(this, {
      expiresAt: estimatedExpiration,
      maxEpoch,
      randomness,
      ephemeralKeyPair: toB64(decodeSuiPrivateKey(ephemeralKeyPair.getSecretKey()).secretKey)
    });
    return oauthUrl;
  }
  // TODO: Should our SDK manage this automatically in addition to exposing a method?
  async handleAuthCallback(hash = window.location.hash) {
    const params = new URLSearchParams(hash.startsWith("#") ? hash.slice(1) : hash);
    const zkp = await this.getSession();
    if (!zkp || !zkp.ephemeralKeyPair || !zkp.maxEpoch || !zkp.randomness) {
      throw new Error(
        "Start of sign-in flow could not be found. Ensure you have started the sign-in flow before calling this."
      );
    }
    const jwt = params.get("id_token");
    if (!jwt) {
      throw new Error("Missing ID Token");
    }
    const decodedJwt = decodeJwt(jwt);
    if (!decodedJwt.sub || !decodedJwt.aud || typeof decodedJwt.aud !== "string") {
      throw new Error("Missing JWT data");
    }
    const { address, salt } = await __privateGet(this, _enokiClient).getZkLogin({ jwt });
    this.$zkLoginState.set({
      ...this.$zkLoginState.get(),
      salt,
      address
    });
    await __privateMethod(this, _setSession, setSession_fn).call(this, {
      ...zkp,
      jwt
    });
    return params.get("state");
  }
  async getSession() {
    if (this.$zkLoginSession.get().initialized) {
      return this.$zkLoginSession.get().value;
    }
    try {
      const storedValue = __privateGet(this, _store).get(__privateGet(this, _storageKeys).SESSION);
      if (!storedValue)
        return null;
      const state = JSON.parse(
        await __privateGet(this, _encryption).decrypt(__privateGet(this, _encryptionKey), storedValue)
      );
      if (state?.expiresAt && Date.now() > state.expiresAt) {
        await this.logout();
      } else {
        this.$zkLoginSession.set({ initialized: true, value: state });
      }
    } catch {
      this.$zkLoginSession.set({ initialized: true, value: null });
    }
    return this.$zkLoginSession.get().value;
  }
  async logout() {
    this.$zkLoginState.set({});
    __privateGet(this, _store).delete(__privateGet(this, _storageKeys).STATE);
    await __privateMethod(this, _setSession, setSession_fn).call(this, null);
  }
  // TODO: Should this return the proof if it already exists?
  async getProof() {
    const zkp = await this.getSession();
    const { salt } = this.$zkLoginState.get();
    if (zkp?.proof) {
      if (zkp.expiresAt && Date.now() > zkp.expiresAt) {
        throw new Error("Stored proof is expired.");
      }
      return zkp.proof;
    }
    if (!salt || !zkp || !zkp.jwt) {
      throw new Error("Missing required parameters for proof generation");
    }
    const ephemeralKeyPair = Ed25519Keypair.fromSecretKey(fromB64(zkp.ephemeralKeyPair));
    const proof = await __privateGet(this, _enokiClient).createZkLoginZkp({
      jwt: zkp.jwt,
      maxEpoch: zkp.maxEpoch,
      randomness: zkp.randomness,
      ephemeralPublicKey: ephemeralKeyPair.getPublicKey()
    });
    await __privateMethod(this, _setSession, setSession_fn).call(this, {
      ...zkp,
      proof
    });
    return proof;
  }
  async getKeypair() {
    await this.getProof();
    const zkp = await this.getSession();
    const { address } = this.$zkLoginState.get();
    if (!address || !zkp || !zkp.proof) {
      throw new Error("Missing required data for keypair generation.");
    }
    if (Date.now() > zkp.expiresAt) {
      throw new Error("Stored proof is expired.");
    }
    return new EnokiKeypair({
      address,
      maxEpoch: zkp.maxEpoch,
      proof: zkp.proof,
      ephemeralKeypair: Ed25519Keypair.fromSecretKey(fromB64(zkp.ephemeralKeyPair))
    });
  }
  async sponsorTransactionBlock({
    network,
    transactionBlock,
    client
  }) {
    const session = await this.getSession();
    if (!session || !session.jwt) {
      throw new Error("Missing required data for sponsorship.");
    }
    const transactionBlockKindBytes = await transactionBlock.build({
      onlyTransactionKind: true,
      client,
      // Theses limits will get verified during the final transaction construction, so we can safely ignore them here:
      limits: {
        maxGasObjects: Infinity,
        maxPureArgumentSize: Infinity,
        maxTxGas: Infinity,
        maxTxSizeBytes: Infinity
      }
    });
    return await __privateGet(this, _enokiClient).createSponsoredTransactionBlock({
      jwt: session.jwt,
      network,
      transactionBlockKindBytes: toB64(transactionBlockKindBytes)
    });
  }
  async executeTransactionBlock({
    bytes,
    digest,
    client
  }) {
    const keypair = await this.getKeypair();
    const userSignature = await keypair.signTransactionBlock(fromB64(bytes));
    await __privateGet(this, _enokiClient).executeSponsoredTransactionBlock({
      digest,
      signature: userSignature.signature
    });
    await client.waitForTransactionBlock({ digest });
    return { digest };
  }
  async sponsorAndExecuteTransactionBlock({
    network,
    transactionBlock,
    client
  }) {
    const { bytes, digest } = await this.sponsorTransactionBlock({
      network,
      transactionBlock,
      client
    });
    return await this.executeTransactionBlock({ bytes, digest, client });
  }
}
_storageKeys = new WeakMap();
_enokiClient = new WeakMap();
_encryption = new WeakMap();
_encryptionKey = new WeakMap();
_store = new WeakMap();
_setSession = new WeakSet();
setSession_fn = async function(newValue) {
  if (newValue) {
    const storedValue = await __privateGet(this, _encryption).encrypt(
      __privateGet(this, _encryptionKey),
      JSON.stringify(newValue)
    );
    __privateGet(this, _store).set(__privateGet(this, _storageKeys).SESSION, storedValue);
  } else {
    __privateGet(this, _store).delete(__privateGet(this, _storageKeys).SESSION);
  }
  this.$zkLoginSession.set({ initialized: true, value: newValue });
};
export {
  EnokiFlow
};
